infrastructures/docker/infrastructure/src/main/java/org/eclipse/che/workspace/infrastructure/docker/ParallelDockerImagesPreparer.java

```diff
@@ -0,0 +1,323 @@
+ * being started.
+ *
+ * @author Max Shaposhnik (mshaposh@redhat.com)
+ */
+public class ParallelDockerImagesPreparer {
+
+  private static final Logger LOG = getLogger(ParallelDockerImagesPreparer.class);
+  private static final String PARALLEL_NUMBER_PROPERTY = "che.infra.docker.max_pull_thread";
+
+  private final RuntimeIdentity identity;
+  private final MachineLoggersFactory machineLoggersFactory;
+  private final boolean doForcePullImage;
+  private final UserSpecificDockerRegistryCredentialsProvider dockerCredentials;
+  private final DockerConnector dockerConnector;
+  private final ThreadPoolExecutor executor;
+
+  @Inject
+  public ParallelDockerImagesPreparer(
+      @Assisted RuntimeIdentity identity,
+      @Named("che.docker.always_pull_image") boolean doForcePullImage,
+      @Named(PARALLEL_NUMBER_PROPERTY) int parallelPullsNumber,
+      UserSpecificDockerRegistryCredentialsProvider dockerCredentials,
+      DockerConnector dockerConnector,
+      MachineLoggersFactory machineLoggersFactory) {
+    this.identity = identity;
+    this.doForcePullImage = doForcePullImage;
+    this.dockerCredentials = dockerCredentials;
+    this.dockerConnector = dockerConnector;
+    this.machineLoggersFactory = machineLoggersFactory;
+
+    ThreadFactory factory =
+        new ThreadFactoryBuilder()
+            .setUncaughtExceptionHandler(LoggingUncaughtExceptionHandler.getInstance())
+            .setNameFormat(getClass().getSimpleName())
+            .build();
+    this.executor = (ThreadPoolExecutor) Executors.newFixedThreadPool(parallelPullsNumber, factory);
+  }
+
+  /**
+   * Schedule parallel preparation of docker images for the set of docker containers.
+   *
+   * @param containers map of machine name and it's container config
+   * @return map of machine names and theirs image names.
+   * @throws InterruptedException if process is interrupted
+   * @throws InfrastructureException if build failed
+   */
+  public Map<String, String> prepareImages(Map<String, DockerContainerConfig> containers)
+      throws InterruptedException, InfrastructureException {
+    if (executor.getActiveCount() + containers.size() > executor.getMaximumPoolSize()) {
+      LOG.warn(
+          String.format(
+              "Maximum parallel images preparing threads reached. Some images will be queued.\n"
+                  + " Workspace machines count is %s. If problem persists, increase %s property value.",
+              containers.size(), PARALLEL_NUMBER_PROPERTY));
+    }
+    Map<String, String> imagesMap = new ConcurrentHashMap<>(containers.size());
+    CompletableFuture<Void> firstFailed = new CompletableFuture<>();
+    List<CompletableFuture<Void>> taskFutures =
+        containers
+            .entrySet()
+            .stream()
+            .map(
+                e ->
+                    CompletableFuture.supplyAsync(
+                        (Supplier<Void>)
```

Comment:
To be honest this method is hard to read. Have you considered moving the lambda to named inner class to simplify this code?

Reply:
I see. I was thinking about passing them to a constructor of a class. But if this doesn't look better for you I'm ok with that.